// frog/_lib/utils/getCastActionContext.js
export function getCastActionContext(parameters) {
    const { context } = parameters;
    const { env, frameData, req, verified } = context || {};
    if (!frameData)
        throw new Error('Frame data must be present for action handlers.');
    return {
        context: {
            actionData: {
                buttonIndex: 1,
                castId: frameData.castId,
                fid: frameData.fid,
                network: frameData.network,
                messageHash: frameData.messageHash,
                timestamp: frameData.timestamp,
                url: frameData.url,
            },
            env,
            error: (data) => ({
                error: data,
                format: 'castAction',
                status: 'error',
            }),
            frame: (data) => ({
                data: { path: data.path, type: 'frame' },
                format: 'castAction',
                status: 'success',
            }),
            message: (data) => ({
                data: { message: data.message, type: 'message' },
                format: 'castAction',
                status: 'success',
            }),
            req,
            res: (data) => ({
                data,
                format: 'castAction',
                status: 'success',
            }),
            var: context.var,
            verified,
        },
    };
}
