// frog/_lib/utils/getSignatureContext.js
import { getIntentState } from './getIntentState.js';
export function getSignatureContext(parameters) {
    const { context } = parameters;
    const { env, frameData, initialPath, previousButtonValues, previousState, req, status, verified, url, } = context || {};
    const { buttonValue, inputText } = getIntentState({
        buttonValues: previousButtonValues || [],
        frameData,
    });
    return {
        context: {
            address: frameData?.address,
            buttonIndex: frameData?.buttonIndex,
            buttonValue,
            env,
            error: (data) => ({
                error: data,
                format: 'signature',
                status: 'error',
            }),
            frameData,
            initialPath,
            inputText,
            previousButtonValues,
            previousState,
            req,
            res(parameters) {
                const { chainId, method, params } = parameters;
                const { domain, types, primaryType, message } = params;
                const response = {
                    chainId,
                    method,
                    params: {
                        domain,
                        types,
                        primaryType,
                        // @TODO: fix typing
                        message: message,
                    },
                };
                return { data: response, format: 'signature', status: 'success' };
            },
            signTypedData(parameters) {
                const { chainId, ...params } = parameters;
                return this.res({
                    params: params,
                    chainId,
                    method: 'eth_signTypedData_v4',
                });
            },
            status,
            var: context.var,
            verified,
            url,
        },
    };
}
